### PR TITLE
Refactor type

### PR DESCRIPTION
**Pull request by HosseinYousefi**
Addresses parts of https://github.com/dart-lang/native/issues/1581
Closes https://github.com/dart-lang/native/issues/664
Addresses parts of https://github.com/dart-lang/native/issues/614

* Made `$type`, `reference`, `fromReference`, inner type params and ... internal
* Made everything in the type classes internal, they are still needed by the other bindings
  * Users now must use `Foo.type` and `GenericFoo.type(...)` to create the type classes instead of `$Foo$Type(...)`.
  * I could alternatively make the type classes internal altogether, but that messes with exporting bindings in a package as users would have to hide them one by one. Alternatively we could generate `.t.dart` files that only contain the typeclasses but that seemed to be overkill for what this realistically does, which is making type class creation syntax consistent.
* Renamed `castTo` to `as` to be closer to Dart casting syntax.
* Relaxed renaming keywords, some identifiers are not possible to use as class names but they are possible to use as method and field names.
* Renamed the type class names from `$FooType` to `$Foo$Type` to prevent name collisions.

### CODE REVIEW

File: pkgs/jni/test/type_test.dart [hunk 1]
```diff
@@ -55,17 +62,22 @@ class B extends JObject {
...
   @override
   B fromReference(JReference reference) {
     return B.fromReference(reference);
   }
 
+  @internal
   @override
   String get signature => 'B';
 
+  @internal
```
Comment (dcharkes) [inline pkgs/jni/test/type_test.dart:75]: duplicate annotations?

**Review by dcharkes — APPROVED** [review]
> Closes https://github.com/dart-lang/native/issues/664
> Addresses parts of https://github.com/dart-lang/native/issues/614

I'd have expected some tests exercising those use cases. Have a test with a method name that would have been shadowed by the previous naming scheme. And have a test that relies on some keyword not be renamed in some context.

LGTM!

**Comment by HosseinYousefi** [comment]
> I'd have expected some tests exercising those use cases. Have a test with a method name that would have been shadowed by the previous naming scheme. And have a test that relies on some keyword not be renamed in some context.

Will be addressed by https://github.com/dart-lang/native/issues/569

File: pkgs/jni/test/type_test.dart (hunk 1 above)
Comment (HosseinYousefi) [reply pkgs/jni/test/type_test.dart:75]: `s/@internal\n\s+@internal/@internal/`!

**Pull request merged**